### PR TITLE
types: rename network type to match Network.config.openshift.io.

### PR DESCRIPTION
**Pull request by squeed**
Renames:
- serviceCIDR -> serviceNetwork
- clusterNetworks -> clusterNetwork
- type -> networkType
- hostSubnetLength -> hostPrefix

Add significantly more network validation, especially around overlapping cidrs.

This adds an upconversion step to loading install config. It will convert any deprecated install configs to the latest version.

### CODE REVIEW

**Comment by squeed** [comment]
cc @abhinavdahiya

**Comment by squeed** [comment]
/retest
nat gateway issue

**Comment by squeed** [comment]
Watching the e2e tests, the cluster came up correctly. Any failure is (probably) a flake.

File: pkg/types/installconfig.go [hunk 1]
```diff
@@ -121,21 +121,35 @@ type Networking struct {
 	// For Libvirt, the default is 192.168.126.0/24.
 	MachineCIDR *ipnet.IPNet `json:"machineCIDR,omitempty"`
```
Comment (squeed) [inline pkg/types/installconfig.go]: Hey @enxebre would you like to add support for multiple MachineCIDRs here too?

File: pkg/types/validation/installconfig_test.go [hunk 2]
```diff
@@ -175,38 +197,59 @@ func TestValidateInstallConfig(t *testing.T) {
...
+					{CIDR: *ipnet.MustParseCIDR("12.0.0.0/16"), HostPrefix: 23},
+					{CIDR: *ipnet.MustParseCIDR("12.0.3.0/24"), HostPrefix: 25},
+				}
 				return c
 			}(),
-			expectedError: `^networking\.clusterNetworks\[0]\.cidr: Invalid value: "172\.30\.2\.0/24": cluster network CIDR must not overlap with serviceCIDR$`,
+			expectedError: `^networking\.clusterNetwork\[1]\.cidr: Invalid value: "12\.0\.3\.0/24": cluster network must not overlap with cluster network 0$`,
 		},
 		{
 			name: "cluster network host subnet length too large",
```
Comment (staebler) [inline pkg/types/validation/installconfig_test.go]: Update name of sub-test.

File: pkg/types/validation/installconfig.go [hunk 3]
```diff
@@ -90,35 +89,56 @@ func validateNetworking(n *types.Networking, fldPath *field.Path) field.ErrorLis
...
 	}
-	if cn.HostSubnetLength < 0 {
-		allErrs = append(allErrs, field.Invalid(fldPath.Child("hostSubnetLength"), cn.HostSubnetLength, "hostSubnetLength must be positive"))
+	if cn.HostPrefix < 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("hostPrefix"), cn.HostPrefix, "hostPrefix must be positive"))
 	}
-	if ones, bits := cn.CIDR.Mask.Size(); cn.HostSubnetLength > int32(bits-ones) {
-		allErrs = append(allErrs, field.Invalid(fldPath.Child("hostSubnetLength"), cn.HostSubnetLength, "cluster network host subnet must not be larger than CIDR "+cn.CIDR.String()))
+	if ones, _ := cn.CIDR.Mask.Size(); cn.HostPrefix < int32(ones) {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("hostPrefix"), cn.HostPrefix, "cluster network host prefix must not be larger than CIDR "+cn.CIDR.String()))
```
Comment (staebler) [inline pkg/types/validation/installconfig.go]: s/larger/smaller/

File: pkg/asset/tls/apiserver.go [hunk 4]
```diff
@@ -34,7 +34,7 @@ func (a *APIServerCertKey) Generate(dependencies asset.Parents) error {
 	installConfig := &installconfig.InstallConfig{}
 	dependencies.Get(kubeCA, installConfig)
 
-	apiServerAddress, err := cidrhost(installConfig.Config.Networking.ServiceCIDR.IPNet, 1)
+	apiServerAddress, err := cidrhost(installConfig.Config.Networking.ServiceNetwork[0].IPNet, 1)
```
[the remark this answers is not in the document]
Comment (squeed) [reply pkg/asset/tls/apiserver.go]: Fordunately no, we don't have to do that. Good question, though.

File: pkg/types/installconfig.go [hunk 5]
```diff
@@ -121,21 +121,35 @@ type Networking struct {
...
 	// Default is OpenShiftSDN.
-	Type string `json:"type,omitempty"`
+	NetworkType string `json:"networkType,omitempty"`
 
-	// ServiceCIDR is the IP address space from which to assign service IPs.
+	// ClusterNetwork is the IP address pool to use for pod IPs.
 	// +optional
-	// Default is 172.30.0.0/16.
-	ServiceCIDR *ipnet.IPNet `json:"serviceCIDR,omitempty"`
+	// Default is 10.128.0.0/14 and a host prefix of /31
```
[the remark this answers is not in the document]
Comment (squeed) [reply pkg/types/installconfig.go]: er, yes. Good catch.

File: pkg/types/installconfig.go [hunk 6]
```diff
@@ -121,21 +121,35 @@ type Networking struct {
...
+	ClusterNetwork []ClusterNetworkEntry `json:"clusterNetwork,omitempty"`
 
-	// ClusterNetworks is the IP address space from which to assign pod IPs.
+	// ServiceNetwork is the ip address pool to use for service IPs.
 	// +optional
-	// Default is a single cluster network with a CIDR of 10.128.0.0/14
-	// and a host subnet length of 9.
-	ClusterNetworks []ClusterNetworkEntry `json:"clusterNetworks,omitempty"`
+	// Default is 172.30.0.0/16
+	// NOTE: currently only one entry is supported.
```
[the remark this answers is not in the document]
Comment (squeed) [reply pkg/types/installconfig.go]: :+1:

File: pkg/types/installconfig.go [hunk 7]
```diff
@@ -144,7 +158,11 @@ type ClusterNetworkEntry struct {
 	// The IP block address pool
 	CIDR ipnet.IPNet `json:"cidr"`
 
+	// HostPrefix is the prefix size to allocate to each node from the CIDR.
+	// For example, 24 would allocate 2^8=256 adresses to each node.
+	HostPrefix int32 `json:"hostPrefix,omitempty"`
```
[the remark this answers is not in the document]
Comment (squeed) [reply pkg/types/installconfig.go]: No, it shouldn't, good point.

File: pkg/types/validation/installconfig.go (hunk 3 above)
Comment (squeed) [reply pkg/types/validation/installconfig.go]: Remember, in networking, 20 is larger than 24 :-)

Comment (staebler) [reply pkg/types/validation/installconfig.go]: I would agree if the message said "cluster network *size* must not be larger". But the message is talking about the host prefix itself not the network size. Would it make sense for the message to be something like "number of hosts must not be larger than number of addresses"?

Comment (staebler) [reply pkg/types/validation/installconfig.go]: But if it makes sense in network parlance as is, then I can live with it.

Comment (squeed) [reply pkg/types/validation/installconfig.go]: yeah, using "size" is a good idea - tweaked the wording a bit.

**Comment by squeed** [comment]
Bumped the version, added some conversion code to bump the version of existing configs, since they're correctly upconverted. PTAL.

File: pkg/types/conversion/installconfig.go [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+	"github.com/openshift/installer/pkg/types"
+	"github.com/pkg/errors"
+)
+
+// ConvertInstallConfig is modeled after the k8s conversion schemes, which is
+// how deprecated values are upconverted.
+// This updates the APIVersion to reflect the fact that we've internally
+// upconverted.
+func ConvertInstallConfig(config *types.InstallConfig) error {
+	// check that the version is convertable
```
Comment (staebler) [inline pkg/types/conversion/installconfig.go]: nit: convertible

Comment (squeed) [reply pkg/types/conversion/installconfig.go]: :man_facepalming:

**Comment by abhinavdahiya** [comment]
the changes specific to changing field names is to create UX consistency with https://github.com/openshift/api/blob/ad60feaa9147fd1ba910f6cbbf3fccee48ac9f17/config/v1/types_network.go#L29-L41 that seems good for 4.0

Also the docs team has been looking for consistency on network related fields like https://bugzilla.redhat.com/show_bug.cgi?id=1684659

and the servicenetwork from string to []string , i'm fine as the validation https://github.com/openshift/installer/pull/1356/files#diff-02200a3eac2ac223b4117e690cf2d0c7R111 seems us in check. 

so /lgtm

**Comment by abhinavdahiya** [comment]
looks like the previous message didn't add lgtm :(

**Pull request merged**